be/src/storage/rowset_update_state.cpp

```diff
@@ -103,6 +109,130 @@ Status RowsetUpdateState::_do_load(Rowset* rowset) {
+            idxes[sort_entries[i].idx] = ridx;
+            ridx++;
+        }
+    }
+}
+
+static void gen_column_by_src_column_and_idxes(const vectorized::Column& src, const vector<uint32_t>& idxes,
+                                               std::unique_ptr<vectorized::Column>* dest) {}
+
+Status RowsetUpdateState::apply(Tablet* tablet, Rowset* rowset, uint32_t rowset_id, PrimaryIndex& index) {
+    if (!rowset->has_txn_meta()) {
+        return Status::OK();
+    }
+    // currently assume it's a partial update
+    const RowsetTxnMetaPB& txn_meta = rowset->rowset_meta()->get_meta_pb().txn_meta();
+    const TabletSchema& tschema = tablet->tablet_schema();
+    // columns supplied in rowset
+    std::vector<uint32_t> update_colum_ids(txn_meta.partial_update_column_ids().begin(),
+                                           txn_meta.partial_update_column_ids().end());
+    std::set<uint32_t> update_columns_set(update_colum_ids.begin(), update_colum_ids.end());
+    // columns needs to be read from tablet's data
+    std::vector<uint32_t> read_column_ids;
+    for (uint32_t i = 0; i < tschema.num_columns(); i++) {
+        if (update_columns_set.find(i) == update_columns_set.end()) {
+            read_column_ids.push_back(i);
+        }
+    }
+    vector<std::unique_ptr<vectorized::Column>> read_columns(read_column_ids.size());
+    vector<std::unique_ptr<vectorized::Column>> write_columns(read_column_ids.size());
+    size_t num_segments = rowset->num_segments();
+    DCHECK(num_segments == _upserts.size());
+    vector<std::pair<string, string>> rewrite_files;
+    DeferOp clean_temp_files([&] {
+        for (auto& e : rewrite_files) {
+            Env::Default()->delete_file(e.second);
+        }
+    });
+    for (size_t i = 0; i < num_segments; i++) {
+        auto& pks = *_upserts[i];
+        int64_t t_start = MonotonicMillis();
+        std::vector<uint64_t> rowids(pks.size());
+        // TODO: get rowid by pk column using primary index
+        int64_t t_get_rowids = MonotonicMillis();
+        size_t num_default = 0;
+        std::map<uint32_t, std::vector<uint32_t>> rowids_by_rssid;
+        vector<uint32_t> idxes;
+        // group rowids by rssid, and for each group sort by rowid
+        plan_read_by_rssid(rowids, &num_default, &rowids_by_rssid, &idxes);
+        // get column values by rowid, also get default values if needed
+        RETURN_IF_ERROR(tablet->updates()->_get_column_values(read_column_ids, num_default > 0, rowids_by_rssid,
+                                                              &read_columns));
+        for (size_t col_idx = 0; col_idx < read_column_ids.size(); i++) {
+            gen_column_by_src_column_and_idxes(*read_columns[col_idx], idxes, &write_columns[col_idx]);
+        }
+        int64_t t_get_column_values = MonotonicMillis();
+        auto src_path = BetaRowset::segment_file_path(tablet->schema_hash_path(), rowset->rowset_id(), i);
+        auto dest_path = src_path + ".tmp";
```

Comment:
can use a specialized function to generate this kind of filename, such as these:
https://github.com/StarRocks/starrocks/blob/9bd6cbb221160fea14ad6c7d3d2acee60d41ed4c/be/src/storage/rowset/beta_rowset.cpp#L47-L61